Microsoft.Toolkit.Uwp.UI.Controls/MasterDetailsView/MasterDetailsView.cs

```diff
@@ -96,25 +96,48 @@ protected override void OnApplyTemplate()
         }
 
         /// <summary>
-        /// Fired when the SelectedItem changes.
+        /// Fired when the SelectedIndex changes.
         /// </summary>
         /// <param name="d">The sender</param>
         /// <param name="e">The event args</param>
         /// <remarks>
         /// Sets up animations for the DetailsPresenter for animating in/out.
         /// </remarks>
-        private static void OnSelectedItemChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        private static void OnSelectedIndexChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
         {
             var view = (MasterDetailsView)d;
 
-            view.OnSelectionChanged(new SelectionChangedEventArgs(new List<object> { e.OldValue }, new List<object> { e.NewValue }));
+            var newValue = (int)e.NewValue == -1 ? null : view.Items[(int)e.NewValue];
+            var oldValue = e.OldValue == null ? null : view.Items.ElementAtOrDefault((int)e.OldValue);
 
-            view.UpdateView(true);
+            // check if selection actually changed
+            if (view.SelectedItem != newValue)
+            {
+                // sync SelectedItem
+                view.SetValue(SelectedItemProperty, newValue);
+                view.UpdateSelection(oldValue, newValue);
+            }
+        }
 
-            // If there is no selection, do not remove the DetailsPresenter content but let it animate out.
-            if (view.SelectedItem != null)
+        /// <summary>
+        /// Fired when the SelectedItem changes.
+        /// </summary>
+        /// <param name="d">The sender</param>
+        /// <param name="e">The event args</param>
+        /// <remarks>
+        /// Sets up animations for the DetailsPresenter for animating in/out.
+        /// </remarks>
+        private static void OnSelectedItemChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        {
+            var view = (MasterDetailsView)d;
+            var index = e.NewValue == null ? -1 : view.Items.IndexOf(e.NewValue);
+
+            // check if selection actually changed
+            if (view.SelectedIndex != index)
             {
-                view.SetDetailsContent();
+                // sync SelectedIndex
+                view.SetValue(SelectedIndexProperty, index);
+                view.UpdateSelection(e.OldValue, e.NewValue);
```

Comment:
>view.UpdateSelection(e.OldValue, e.NewValue); [](start = 16, length = 45)

This will already be executed in `OnSelectedIndexChanged()` after calling `view.SetValue(SelectedIndexProperty, index);`. You should be able to remove this call from here.